homeassistant/components/pvpc_hourly_pricing/config_flow.py

```diff
@@ -33,36 +60,168 @@ async def async_step_user(
+            self._tariff = user_input[ATTR_TARIFF]
+            self._power = user_input[ATTR_POWER]
+            self._power_p3 = user_input[ATTR_POWER_P3]
+            self._use_api_token = user_input[CONF_USE_API_TOKEN]
+            return self.async_show_form(
+                step_id="api_token",
+                data_schema=vol.Schema(
+                    {vol.Required(CONF_API_TOKEN, default=self._api_token): str}
+                ),
+                description_placeholders={"mail_to_link": _MAIL_TO_LINK},
+            )
+
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_NAME, default=DEFAULT_NAME): str,
+                vol.Required(ATTR_TARIFF, default=DEFAULT_TARIFF): VALID_TARIFF,
+                vol.Required(ATTR_POWER, default=DEFAULT_POWER_KW): VALID_POWER,
+                vol.Required(ATTR_POWER_P3, default=DEFAULT_POWER_KW): VALID_POWER,
+                vol.Required(CONF_USE_API_TOKEN, default=False): bool,
+            }
+        )
+        return self.async_show_form(step_id="user", data_schema=data_schema)
+
+    async def async_step_api_token(self, user_input: dict[str, Any]) -> FlowResult:
+        """Handle optional step to define API token for extra sensors."""
+        self._api_token = user_input[CONF_API_TOKEN]
+        return await self._async_verify(
+            "api_token",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_API_TOKEN, default=self._api_token): str}
+            ),
+        )
+
+    async def _async_verify(self, step_id: str, data_schema: vol.Schema) -> FlowResult:
+        """Attempt to verify the provided configuration."""
+        data = {
+            CONF_NAME: self._name,
+            ATTR_TARIFF: self._tariff,
+            ATTR_POWER: self._power,
+            ATTR_POWER_P3: self._power_p3,
+            CONF_USE_API_TOKEN: self._use_api_token,
+            CONF_API_TOKEN: self._api_token,
+        }
+        errors: dict[str, str] = {}
+        auth_ok = True
+        if self._use_api_token:
+            if not self._api:
+                self._api = PVPCData(session=async_get_clientsession(self.hass))
+            auth_ok = await self._api.check_api_token(dt_util.utcnow(), self._api_token)
+        if not auth_ok:
+            errors["base"] = "invalid_auth"
+            return self.async_show_form(
+                step_id=step_id,
+                data_schema=data_schema,
+                errors=errors,
+                description_placeholders={"mail_to_link": _MAIL_TO_LINK},
+            )
 
-        return self.async_show_form(step_id="user", data_schema=UI_CONFIG_SCHEMA)
+        if self._reauth_entry:
+            self.hass.config_entries.async_update_entry(self._reauth_entry, data=data)
+            self.hass.async_create_task(
+                self.hass.config_entries.async_reload(self._reauth_entry.entry_id)
+            )
+            return self.async_abort(reason="reauth_successful")
```

Comment:
Why not just let this continue through and end with the `self.async_create_entry` return value below? Why end in an abort option rather than a successful return? I know the abort return is providing a success message but it still feels odd to abort here rather than end is success.

Reply:
Whenever I need a pattern I look to ZHA and Zwave since they are maintained by founders. In those integrations, I see that any time `async_update_entry` is called the return is `async_create_entry`. So the recommendation would be:
```
        if self._reauth_entry:
            self.hass.config_entries.async_update_entry(self._reauth_entry, data=data)
            self.hass.async_create_task(
                self.hass.config_entries.async_reload(self._reauth_entry.entry_id)
            )
        assert self._name is not None
        return self.async_create_entry(title=self._name, data=data)
```
This will result in a successful return for the user. We can ask Martin as he is very good with this stuff:
@MartinHjelmare do you have an opinion here. It feels to me that an `async_abort` return on a successful result is not the intention vs `async_create_entry` which feels redundant to `async_update_entry`.

Reply:
I wouldn't look at zha or zwave_js. Those integrations have very complex flows that control add-ons etc.

The standard is to abort the flow with `reauth_successful` reason after re-authentication.

Reply:
> The standard is to abort the flow with reauth_successful reason after re-authentication.

Then it's ok?

Another (related) doubt I had around the 'reauth' and that code snippet was about **waiting the reload or just firing a task for it**... 

I've also seen it both ways, and it was not clear for me how to act... 

Finally, I fired the `async_create_task` for the reload, instead of awaiting it (because it's not necessary at that point to double-check), and had to put an extra `hass.async_block_till_done()` in the end of `test_reauth` , between `call_count=6` (the successful verification in the reauth flow) and `call_count=7` (the final call in the entry reload).

But now I see 101 matches for `await self.hass.config_entries.async_reload()` for a total 167 matches of `self.hass.config_entries.async_reload()` 🙈, so it looks like the usual thing is to await the reload, right?

**Should I change it?**

Reply:
What I am seeing from the documentation provided is that your approach is correct. I will approve.